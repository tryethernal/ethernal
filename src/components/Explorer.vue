<template>
    <v-container fluid>
        <template v-if="justCreated">
            <v-card outlined>
                <v-card-text>
                    <v-progress-circular
                        class="mr-2"
                        indeterminate
                        size="16"
                        width="2"
                        color="primary"></v-progress-circular>
                    Your subscripion is being activated. The explorer will be ready soon.
                </v-card-text>
            </v-card>
        </template>
        <template v-else-if="explorer && workspaces.length > 0">
            <v-alert text type="warning" v-if="explorer.stripeSubscription && explorer.stripeSubscription.isTrialing">
                This explorer is on a free trial plan. To keep it running once it's over, add a payment method.
            </v-alert>
            <v-alert text type="error" v-if="!explorer.stripeSubscription">This explorer is not active. To activate it, start a subscription.</v-alert>
            <h2>{{ explorer.name }}</h2>
            <v-row>
                <v-col cols="6">
                    <h4>Settings</h4>
                    <Explorer-Settings :key="JSON.stringify(capabilities)" :explorer="explorer" :workspaces="workspaces" @updated="loadExplorer(id)"/>
                </v-col>
                <v-col cols="6">
                    <template v-if="isBillingEnabled">
                        <h4>Billing</h4>
                        <Explorer-Billing :explorer="explorer" @updated="loadExplorer(id)" />
                    </template>
                    <h4 :class="{ 'mt-2': isBillingEnabled }">Domain Aliases</h4>
                    <Explorer-Domains-List :key="JSON.stringify(capabilities)" :explorer="explorer" :disabled="isBillingEnabled && (!explorer.stripeSubscription || !explorer.stripeSubscription.stripePlan.capabilities.customDomain)" @updated="loadExplorer(id)" />
                </v-col>
            </v-row>
            <v-row>
                <v-col>
                    <h4>Branding</h4>
                    <Explorer-Branding :key="JSON.stringify(capabilities)" :explorer="explorer" :disabled="isBillingEnabled && (!explorer.stripeSubscription || !explorer.stripeSubscription.stripePlan.capabilities.branding)" @updated="loadExplorer(id)" />
                </v-col>
            </v-row>
            <v-row>
                <v-col cols="6">
                    <h4 class="error--text">Danger Zone</h4>
                    <Explorer-Danger-Zone :key="JSON.stringify(capabilities)" :explorer="explorer" />
                </v-col>
            </v-row>
        </template>
        <template v-else>
            <v-skeleton-loader type="list-item-three-line"></v-skeleton-loader>
        </template>
    </v-container>
</template>

<script>
import { mapGetters } from 'vuex';
import ExplorerSettings from './ExplorerSettings';
import ExplorerBilling from './ExplorerBilling';
import ExplorerDomainsList from './ExplorerDomainsList';
import ExplorerBranding from './ExplorerBranding';
import ExplorerDangerZone from './ExplorerDangerZone';

export default {
    name: 'Explorer',
    props: ['id'],
    components: {
        ExplorerSettings,
        ExplorerBilling,
        ExplorerDomainsList,
        ExplorerBranding,
        ExplorerDangerZone
    },
    data: () => ({
        workspaces: [],
        explorer: null,
        capabilities: {},
        refreshInterval: null
    }),
    methods: {
        loadExplorer(id) {
            this.server.getWorkspaces()
                .then(({ data }) => this.workspaces = data)
                .catch(console.log);

            this.server.getExplorer(id)
                .then(({ data }) => {
                    this.explorer = data;
                    if (this.explorer.stripeSubscription) {
                        if (this.refreshInterval)
                            clearInterval(this.refreshInterval);
                        this.capabilities = this.explorer.stripeSubscription.stripePlan.capabilities;
                    }
                })
                .catch(console.log);
        }
    },
    computed: {
        ...mapGetters([
            'isBillingEnabled'
        ]),
        justCreated() {
            if (!this.explorer)
                return false;
            return !this.explorer.stripeSubscription && this.$route.query.status == 'success';
        }
    },
    watch: {
        id: {
            immediate: true,
            handler(id) {
                this.loadExplorer(id);
                if (this.$route.query.status == 'success')
                    this.refreshInterval = setInterval(this.loadExplorer, 3000, id);
            }
        }
    }
}
</script>
