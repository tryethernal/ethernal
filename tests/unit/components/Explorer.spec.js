import flushPromises from 'flush-promises';
import MockHelper from '../MockHelper';

import Explorer from '@/components/Explorer.vue';

beforeEach(() => jest.clearAllMocks());

describe('Explorer.vue', () => {
    let helper;

    beforeEach(() => {
        helper = new MockHelper();
    });

    it('Should display inactive subscription message', async() => {
        jest.spyOn(helper.mocks.server, 'getWorkspaces').mockResolvedValueOnce({ data: [{ id: 1 }]});
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: { id: 1 }});
        const wrapper = helper.mountFn(Explorer, {
            stubs: ['Explorer-Settings', 'Explorer-Billing', 'Explorer-Domains-List', 'Explorer-Branding', 'Explorer-Danger-Zone']
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display trial message', async() => {
        jest.spyOn(helper.mocks.server, 'getWorkspaces').mockResolvedValueOnce({ data: [{ id: 1 }]});
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: { id: 1, stripeSubscription: { isTrialing: true, stripePlan: { capabilities: {}}}}});
        const wrapper = helper.mountFn(Explorer, {
            stubs: ['Explorer-Settings', 'Explorer-Billing', 'Explorer-Domains-List', 'Explorer-Branding', 'Explorer-Danger-Zone']
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display the explorer page', async () => {
        jest.spyOn(helper.mocks.server, 'getWorkspaces').mockResolvedValueOnce({ data: [{ id: 1 }]});
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: { id: 1, stripeSubscription: { stripePlan: { capabilities: {}}}}});
        const wrapper = helper.mountFn(Explorer, {
            stubs: ['Explorer-Settings', 'Explorer-Billing', 'Explorer-Domains-List', 'Explorer-Branding', 'Explorer-Danger-Zone']
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should display loading', async () => {
        jest.spyOn(helper.mocks.server, 'getWorkspaces').mockResolvedValueOnce({ data: [{ id: 1 }]});
        jest.spyOn(helper.mocks.server, 'getExplorer').mockResolvedValueOnce({ data: { id: 1 }});
        const wrapper = helper.mountFn(Explorer, {
            stubs: ['Explorer-Settings', 'Explorer-Billing', 'Explorer-Domains-List', 'Explorer-Branding', 'Explorer-Danger-Zone'],
            computed: {
                justCreated() { return true }
            }
        });

        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });
});
