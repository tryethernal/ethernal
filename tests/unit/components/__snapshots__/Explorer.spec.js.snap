// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explorer.vue Should display inactive subscription message 1`] = `
<div class="container container--fluid">
  <!---->
  <div role="alert" class="v-alert v-sheet theme--dark v-alert--text error--text">
    <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
      <div class="v-alert__content">This explorer is not active. To activate it, start a subscription.</div>
    </div>
  </div>
  <h2>
    <span class="text-caption">- (<a href="//test.tryethernal.com" target="_blank">test.tryethernal.com</a>)</span>
  </h2>
  <div class="row mt-2">
    <div class="col col-6">
      <h4>Settings</h4>
      <explorer-settings-stub workspaces="[object Object]" explorer="[object Object]"></explorer-settings-stub>
    </div>
    <div class="col col-6">
      <h4>Billing</h4>
      <explorer-billing-stub explorer="[object Object]"></explorer-billing-stub>
      <h4 class="mt-2">Domain Aliases</h4>
      <explorer-domains-list-stub disabled="true" explorer="[object Object]"></explorer-domains-list-stub>
    </div>
  </div>
  <div class="row">
    <div class="col">
      <h4>Branding</h4>
      <explorer-branding-stub disabled="true" explorer="[object Object]"></explorer-branding-stub>
    </div>
  </div>
  <div class="row">
    <div class="col col-6">
      <h4 class="error--text">Danger Zone</h4>
      <explorer-danger-zone-stub explorer="[object Object]"></explorer-danger-zone-stub>
    </div>
  </div>
</div>
`;

exports[`Explorer.vue Should display loading 1`] = `
<div class="container container--fluid">
  <div class="v-card v-sheet v-sheet--outlined theme--light">
    <div class="v-card__text">
      <div role="progressbar" aria-valuemin="0" aria-valuemax="100" class="v-progress-circular mr-2 v-progress-circular--visible v-progress-circular--indeterminate primary--text" style="height: 16px; width: 16px;"><svg xmlns="http://www.w3.org/2000/svg" viewBox="22.857142857142858 22.857142857142858 45.714285714285715 45.714285714285715" style="transform: rotate(0deg);">
          <circle fill="transparent" cx="45.714285714285715" cy="45.714285714285715" r="20" stroke-width="5.714285714285714" stroke-dasharray="125.664" stroke-dashoffset="125.66370614359172px" class="v-progress-circular__overlay"></circle>
        </svg>
        <div class="v-progress-circular__info"></div>
      </div>
      Your subscripion is being activated. The explorer will be ready soon.
    </div>
  </div>
</div>
`;

exports[`Explorer.vue Should display the explorer page 1`] = `
<div class="container container--fluid">
  <!---->
  <!---->
  <h2>
    <span class="text-caption">- (<a href="//a.test.com" target="_blank">a.test.com</a>)</span>
  </h2>
  <div class="row mt-2">
    <div class="col col-6">
      <h4>Settings</h4>
      <explorer-settings-stub workspaces="[object Object]" explorer="[object Object]"></explorer-settings-stub>
    </div>
    <div class="col col-6">
      <h4>Billing</h4>
      <explorer-billing-stub explorer="[object Object]"></explorer-billing-stub>
      <h4 class="mt-2">Domain Aliases</h4>
      <explorer-domains-list-stub disabled="true" explorer="[object Object]"></explorer-domains-list-stub>
    </div>
  </div>
  <div class="row">
    <div class="col">
      <h4>Branding</h4>
      <explorer-branding-stub disabled="true" explorer="[object Object]"></explorer-branding-stub>
    </div>
  </div>
  <div class="row">
    <div class="col col-6">
      <h4 class="error--text">Danger Zone</h4>
      <explorer-danger-zone-stub explorer="[object Object]"></explorer-danger-zone-stub>
    </div>
  </div>
</div>
`;

exports[`Explorer.vue Should display trial message 1`] = `
<div class="container container--fluid">
  <div role="alert" class="v-alert v-sheet theme--dark v-alert--text warning--text">
    <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-exclamation theme--dark warning--text"></i>
      <div class="v-alert__content">
        This explorer is on a free trial plan. To keep it running once it's over, add a payment method.
      </div>
    </div>
  </div>
  <!---->
  <h2>
    <span class="text-caption">- (<a href="//test.tryethernal.com" target="_blank">test.tryethernal.com</a>)</span>
  </h2>
  <div class="row mt-2">
    <div class="col col-6">
      <h4>Settings</h4>
      <explorer-settings-stub workspaces="[object Object]" explorer="[object Object]"></explorer-settings-stub>
    </div>
    <div class="col col-6">
      <h4>Billing</h4>
      <explorer-billing-stub explorer="[object Object]"></explorer-billing-stub>
      <h4 class="mt-2">Domain Aliases</h4>
      <explorer-domains-list-stub disabled="true" explorer="[object Object]"></explorer-domains-list-stub>
    </div>
  </div>
  <div class="row">
    <div class="col">
      <h4>Branding</h4>
      <explorer-branding-stub disabled="true" explorer="[object Object]"></explorer-branding-stub>
    </div>
  </div>
  <div class="row">
    <div class="col col-6">
      <h4 class="error--text">Danger Zone</h4>
      <explorer-danger-zone-stub explorer="[object Object]"></explorer-danger-zone-stub>
    </div>
  </div>
</div>
`;
